### PR TITLE
Clarify event timestamp origin and range

### LINKED ISSUE #798: Define if user-defined timestamp for events must be between span start and end or not
Related spec section:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#add-events

We should specify if the user-provided timestamp has to be between start and end of the span (inclusive) or not. Otherwise it will likely mess up analysis and presentation on some backends since an out-of-range timestamp might be unexpected. If it's out of this range, it should be normalized to either start or end, whichever is closer.

If there are valid reasons for the event timestamp to be before the span started or after it ended, however, this should also be clearly specified in the spec to make potential consumers aware of this and allow proper handling.

### PR DESCRIPTION
**Pull request by arminru**
Resolves #798

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,7 @@ Updates:
   * SpanProcessors must provide read/write access at least in OnStart.
 - Specify how `Probability` sampler is used with `ParentOrElse` sampler.
 - Clean up api-propagators.md, by extending documentation and removing redundant sections ([#577](https://github.com/open-telemetry/opentelemetry-specification/pull/577))
+- Document that event timestamps might be out of range ([#839](https://github.com/open-telemetry/opentelemetry-specification/pull/839))
```
Comment (Oberon00) [inline CHANGELOG.md:25]: Does this need a changelog entry?

File: specification/trace/api.md [hunk 2]
```diff
@@ -448,6 +448,12 @@ The Span interface MUST provide:
 Events SHOULD preserve the order in which they're set. This will typically match
 the ordering of the events' timestamps.
```
Comment (Oberon00) [inline specification/trace/api.md:449]: Regarding this, because I think it's related: Events are not "set" but recorded. But I would note:

```suggestion
Events SHOULD preserve the order in which they are recorded.
This will typically match the ordering of the events' timestamps,
but events may be recorded out-of-order using custom timestamps.
```

**Review by Oberon00 — COMMENTED** [review]
I think we should also add a recommendation for API users to not use custom timestamps unless required.

**Review by Oberon00 — APPROVED** [review]

**Review by iNikem — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (arminru) [reply CHANGELOG.md:25]: It's mainly to make consumers aware that they might encounter unexpected timestamps.

**Comment by arminru** [comment]
@Oberon00 
> I think we should also add a recommendation for API users to not use custom timestamps unless required.

I would've expected this to be added as part of #796/#824 but apparently this was overlooked so I'll add it here.

**Review by carlosalberto — APPROVED** [review]

File: specification/trace/api.md [hunk 3]
```diff
@@ -429,7 +429,10 @@ An `Event` is defined by the following properties:
 
 - (Required) Name of the event.
 - (Optional) [`Attributes`](../common/common.md#attributes).
-- (Optional) Timestamp for the event. If not provided, the current time when the event is added MUST be used.
+- (Optional) Custom Timestamp for the event. A custom timestamp SHOULD only be provided,
+if there is a significant difference between the time at which the event occurred and
```
Comment (tigrannajaryan) [inline specification/trace/api.md:433]: >if there is a significant difference

What constitutes "significant"? Why do we care here in the spec?  If the developer chooses to set a custom timestamp are we in a position to tell them it is not significant?

IMO, if the developer decides to provide the timestamp then that's developer's business, not spec's business.

Perhaps there is something important I am missing here, please correct me if I am wrong.

File: specification/trace/api.md [hunk 4]
```diff
@@ -445,8 +448,15 @@ The Span interface MUST provide:
...
   providing a wrapping class or function it SHOULD be named `EventFormatter`.
 
-Events SHOULD preserve the order in which they're set. This will typically match
-the ordering of the events' timestamps.
+Events SHOULD preserve the order in which they are recorded.
+This will typically match the ordering of the events' timestamps,
+but events may be recorded out-of-order using custom timestamps.
+
+Consumers should be aware that an event's timestamp might be before the start or
+after the end of the span if such timestamp was provided by the user to the
```
Comment (tigrannajaryan) [inline specification/trace/api.md:456]: >if such timestamp was provided by the user to the `AddEvent` API.

 It can also happen if a custom timestamp is provided to Span Start/End function.

File: specification/trace/api.md (hunk 3 above)
Comment (arminru) [reply specification/trace/api.md:433]: I actually meant *significant as per the developer's discretion*. This should emphasize that the default case would be having it added automatically by the SDK's clock and that custom timestamps are only required if one intends to override that one for whatever reason. Do you think the previous wording is sufficient here?

File: specification/trace/api.md (hunk 4 above)
Comment (arminru) [reply specification/trace/api.md:456]: I thought that if one would specify custom timestamps for recording past (or future 😅) spans they would also specify custom timestamps for events but you're right, they could rely on the automatic timestamps at `AddEvent` time here and provoke out-of-range timestamps this way. Will adapt the wording.

File: specification/trace/api.md (hunk 3 above)
Comment (tigrannajaryan) [reply specification/trace/api.md:433]: I see now. I think the way it is worded does not convey the intent as you described in the comment.
I would vote for deleting the sentence. I think it is implied that an "optional" API functionality is there for the developer to use if they wish so, if it is "significant" for them. There is no need to explicitly call it out.

File: specification/trace/api.md [hunk 5]
```diff
@@ -445,8 +448,15 @@ The Span interface MUST provide:
...
-the ordering of the events' timestamps.
+Events SHOULD preserve the order in which they are recorded.
+This will typically match the ordering of the events' timestamps,
+but events may be recorded out-of-order using custom timestamps.
+
+Consumers should be aware that an event's timestamp might be before the start or
+after the end of the span if custom timestamps were provided by the user for the
+event or when starting or ending the span.
+The specification does not require any normalization if provided timestamps are
+out of range.
```
Comment (tigrannajaryan) [inline specification/trace/api.md:459]: 👍

File: specification/trace/api.md (hunk 3 above)
Comment (arminru) [reply specification/trace/api.md]: Actually, from an exporter's and subsequent consumer's POV, a timestamp will *always* be there. I thus adapted the "an event is defined by" section accordingly and specified that a custom timestamp can be provided to the `AddEvent` API which is more precise. PTAL at d5ba3bb.

File: specification/trace/api.md [hunk 6]
```diff
@@ -428,15 +428,20 @@ with the moment when they are added to the `Span`.
 An `Event` is defined by the following properties:
 
 - (Required) Name of the event.
+- (Required) A timestamp for the event. Either the time at which the event was
```
Comment (bogdandrutu) [inline specification/trace/api.md:431]: What? This is not required.

Comment (arminru) [reply specification/trace/api.md:431]: See https://github.com/open-telemetry/opentelemetry-specification/pull/839#discussion_r474031058.

A timestamp is always there from an exporter's and subsequent consumer's POV. An event is defined by these properties and since it's always there it can be seen as a required part of an event.
I clarified further below that the timestamp passed to `AddEvent` is optional and used to override the timestamp which would otherwise be automatically set by the implementation. There is, however, no case in which there would be no timestamp for an event.

Comment (arminru) [reply specification/trace/api.md:431]: See https://github.com/open-telemetry/opentelemetry-specification/pull/839#discussion_r474031058.

A timestamp is always there from an exporter's and subsequent consumer's POV. An event is defined by these properties and since it's always there it can be seen as a required part of an event.
I clarified further below that the timestamp passed to `AddEvent` is optional and used to override the timestamp which would otherwise be automatically set by the implementation. There is, however, no case in which there would be no timestamp for an event and therefore it's not optional.

Comment (bogdandrutu) [reply specification/trace/api.md:431]: The first consumer of the API is not the exporter/vendor is the dev that instruments an app, so I think the specs should be written with that in mind.

Having an optional component with a default value is equivalent for the exporter to be "required".

My concern of calling this required stands, so will not approve this change.

Comment (tigrannajaryan) [reply specification/trace/api.md:431]: @arminru I agree. This a mandatory property of the Event object. It is an optional parameter of the API call.

Comment (bogdandrutu) [reply specification/trace/api.md:431]: @tigrannajaryan sure, but the main consumer of the API is the instrumentation dev, which should not read that it is required to pass a `Timestamp` and as I suggested having an optional field with a default value is equivalent to always present on the consume side.

Comment (tigrannajaryan) [reply specification/trace/api.md:431]: @bogdandrutu you are right, depending on who reads this the information may need to communicated in different ways.

For OpenTelemetry language library authors it must be mandatory to have a timestamp in the Event object and ensure it is always recorded (either automatically or via an API call) and transmitted when exporting.

For API users it should be clear that is not mandatory for them to specify the timestamp and if they don't the default (current) value will be used.

I am not sure what exactly is the right way to express this in the spec. I believe @arminru wanted to make this clear for language library authors.

Comment (bogdandrutu) [reply specification/trace/api.md:431]: 100% agree. So I still believe that we should make sure if the API users reads this will not do it, and it is likely they will read. I would sacrifice and make library authors and exporter authors read more stuff like extra notes than having this "required" here.

Comment (bogdandrutu) [reply specification/trace/api.md:431]: Also we have optional time for start, there as well will ensure it is available (also to make it clear is the SDK responsibility to ensure that not the API)

Comment (bogdandrutu) [reply specification/trace/api.md:431]: We cannot talk about exporters in the context of the API that is something available in the SDK and most likely will confuse readers

Comment (tigrannajaryan) [reply specification/trace/api.md:431]: OK, let's make sure we use the wording "optional" and "required" consistently in the spec. For example in the "Span" description we have this:
```
    The span name
    An immutable SpanContext that uniquely identifies the Span
    A parent span in the form of a Span, SpanContext, or null
    A SpanKind
    A start timestamp
    An end timestamp
    Attributes
    A list of Links to other Spans
    A list of timestamped Events
    A Status.
```
It does not say anywhere that start and end timestamps are optional, but they are optional in the API. Neither does it say "required".
Let's choose one approach and use it consistently.

Comment (bogdandrutu) [reply specification/trace/api.md:431]: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#span-creation
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#end

Comment (bogdandrutu) [reply specification/trace/api.md:431]: For end it is clearly stated as optional, for span creation it uses a different wording, and it says required only for the name

Comment (tigrannajaryan) [reply specification/trace/api.md:431]: @arminru please feel free to either align the wording to match the rest of the spec or propose a future PR which possibly goes over all places where we use "optional"/"required" qualifiers and make sure it is uniform.

Perhaps to make it easier to merge this PR you can just drop this particular change, since I believe your other clarifications later in the document regarding timestamp orders, ranges, etc are important and we can merge them.

Comment (bogdandrutu) [reply specification/trace/api.md:431]: Strongly encourage to revert that change, and if anyone feels the need to change everywhere the wording then we should do that in a separate PR.

Comment (arminru) [reply specification/trace/api.md]: I removed the (Required)/(Optional) prefix here since both would be confusing because the timestamp is not an optional property of an event but optional as an argument to the API call defined below. I filed #850 for having the notation unified across the whole file.

**Review by Oberon00 — APPROVED** [review]
LGTM! I think these changes will also be helpful as preparation for resolving #814.

File: specification/trace/api.md [hunk 7]
```diff
@@ -427,16 +427,21 @@ with the moment when they are added to the `Span`.
 
 An `Event` is defined by the following properties:
 
-- (Required) Name of the event.
-- (Optional) [`Attributes`](../common/common.md#attributes).
-- (Optional) Timestamp for the event. If not provided, the current time when the event is added MUST be used.
+- Name of the event.
+- A timestamp for the event. Either the time at which the event was
+added or a custom timestamp provided by the user.
+- [`Attributes`](../common/common.md#attributes) further describing the event.
```
Comment (bogdandrutu) [inline specification/trace/api.md:433]: As mentioned before, I think this is an unrelated change and should be made across the entire file if we want.

Comment (arminru) [reply specification/trace/api.md:426]: Keeping the wording that the timestamp is an optional property of an event is wrong, however, so dropping the misleading part and opening #850 for consolidating it across the file was the best compromise to move forward IMHO.

This section describes the event itself with its non-optional timestamp and the API definition for _adding_ events below explicitly states the timestamp is optional _as a parameter to this API_ and that it will use the current time as a default if not provided.

We only have these (Required)/(Optional) prefixes for the Events and Links sections here and now only Links are left with it. All the other API definitions don't use this notion anyway so it's not aligned already and this change is not making anything worse. I also don't think the change is unrelated to this PR which is intended to clarify the timestamps origin and range.

Comment (arminru) [reply specification/trace/api.md:440]: @bogdandrutu ^^^
I think we should be able to move forward with this PR and sort out the rest on #850.

Comment (arminru) [reply specification/trace/api.md:438]: @bogdandrutu ^^^
I think we should be able to move forward with this PR and sort out the rest on #850.

**Comment by Oberon00** [comment]
It seems to me that TC members are 2:1 for this PR. But maybe this is not important enough to make a vote? @open-telemetry/technical-committee @bogdandrutu either vote whether we should include this PR or decide that @bogdandrutu's opposition is accepted and the PR should be closed.

**Comment by bogdandrutu** [comment]
@Oberon00 We are not a democracy :)) so far we decided to respect "no" from any TC members, so we never use votes like that.

I am happy to approve this if @arminru or @Oberon00 promise to go and fix the content for links and TraceProvider to be consistent with this.

**Comment by Oberon00** [comment]
So maybe it would be better to fix the consistency *before* this PR?

**Comment by arminru** [comment]
@bogdandrutu I think @Oberon00's comment was primarily intended to indicate this PR is stuck after changes were suggested but despite these comments being addressed in August, no reply was given for more than three weeks and thus blocking the PR.

Sure, that's what I created #850 for and I just assigned it to me.

**Review by bogdandrutu — APPROVED** [review]

**Comment by bogdandrutu** [comment]
@arminru you have the power, you can merge this :)
